File: QID-3223-SFEDeltahedgingLogic/SFEDeltahedgingLogic.R

rm(list = ls(all = TRUE))
graphics.off()

# parameter settings
n   = 50                    # periods (steps)
K   = 100                   # exercise price
S0  = 98                    # initial stock price
sig = 0.2                   # volatility (uniform distributed on 0.1 to 0.5)
r   = 0.05                  # interest rate (uniform distributed on 0 to 0.1)
t0  = 6/52                  # current time (1 week = 1/52)
T   = 26/52                 # maturity
dt  = (T - t0)/n            # period between steps n 
t   = seq(t0, T, l = n)     # T-t0 divided in n intervals
tau = T - t                 # time to maturity

# Standard Wiener Process for path of the stock price S
Wt1	= c(0, sqrt(dt) * cumsum(rnorm(n - 1, 0, 1)))
S1  = S0 * exp((r - 0.5 * sig^2) * t + sig * Wt1)

# 1st path
y1     = (log(S1/K) + (r - sig^2/2) * tau)/(sig * sqrt(tau))
delta1 = pnorm(y1 + sig * sqrt(tau))

# 2nd path
Wt2    = c(0, sqrt(dt) * cumsum(rnorm(n - 1, 0, 1)))
S2     = S0 * exp((r - 0.5 * sig^2) * t + sig * Wt2)
y2     = (log(S2/K) + (r - sig^2/2) * tau)/(sig * sqrt(tau))
delta2 = pnorm(y2 + sig * sqrt(tau))

# 3 Plots
par(mfrow = c(3, 1))

# Plot S(t) vs. tau
plot(S1 ~ seq(1, n), main = "Stock Price Paths", xlab = "Steps", ylab = "Stock price", 
    ylim = c(min(S1, S2), max(S1, S2)), type = "l", col = "red")
lines(S2 ~ seq(1, n), col = "blue")
abline(h = K, col = 1)

# Plot Delta vs. tau
plot(delta1 ~ seq(1, n), main = "Dependence of Delta on Steps", xlab = "Steps", 
    ylab = "Delta", ylim = c(0, 1), type = "o", col = "red")
lines(delta2 ~ seq(1, n), type = "o", col = "blue")

# Plot Delta vs. S(t)
plot(delta1 ~ S1, main = "Dependence of Delta on Stock Price", xlab = "Stock price", 
    ylab = "Delta", xlim = c(min(S1, S2), max(S1, S2)), ylim = c(0, 1), type = "p", 
    col = "red")
points(delta2 ~ S2, col = "blue")
abline(lm(delta1 ~ S1), col = "red")
abline(lm(delta2 ~ S2), col = "blue")

File: QID-3223-SFEDeltahedgingLogic/SFEDeltaHedgingLogic.py
import numpy as np
from scipy.stats import norm
from matplotlib import pyplot as plt

# parameter settings
save_figure = True
n = 50                          # periods (steps)
K = 100                         # exercise price
S0 = 98                         # initial stock price
sig = 0.2                       # volatility
r = 0.05                        # interest rate
t0 = 6/52                       # current time (1 week = 1/52)
T = 26/52                       # maturity
dt = (T - t0)/n                 # period between steps n
t = np.linspace(t0, T, n)       # T-t0 divided in n intervals
tau = T - t                     # time to maturity

# Standard Wiener Process for path of the stock price S
Wt1 = np.concatenate((np.array([0]), np.sqrt(dt)*np.cumsum(np.random.normal(size=n-1))))
S1 = S0 * np.exp((r - 0.5 * sig**2) * t + sig * Wt1)

# 1st path
with np.errstate(divide='ignore'):
    y1 = (np.log(S1/K) + (r - sig**2/2) * tau)/(sig * np.sqrt(tau))
delta1 = norm.cdf(y1 + sig * np.sqrt(tau))

# 2nd path
Wt2 = np.concatenate((np.zeros(1), np.sqrt(dt)*np.cumsum(np.random.normal(size=n-1))))
S2 = S0 * np.exp((r - 0.5 * sig**2) * t + sig * Wt2)
with np.errstate(divide='ignore'):
    y1 = (np.log(S2/K) + (r - sig**2/2) * tau)/(sig * np.sqrt(tau))
delta2 = norm.cdf(y1 + sig * np.sqrt(tau))

# 3 Plots
fig = plt.figure()

# Plot S(t) vs. tau
ax = fig.add_subplot(3,1,1)
ax.plot(S1, c='r')
ax.plot(S2, c='b')
ax.axhline(K, c='black', linestyle='dashed')
ax.set_title('Stock Price Paths')
ax.set_xlabel('Steps')
ax.set_ylabel('Stock price')
ax.legend(('S1', 'S2', 'K'))

# Plot Delta vs. tau
ax = fig.add_subplot(3,1,2)
ax.plot(delta1, c='r')
ax.plot(delta2, c='b')
ax.set_title('Dependence of Delta on Steps')
ax.set_xlabel('Steps')
ax.set_ylabel('Delta')
ax.legend(('Delta 1', 'Delta 2'))

# Plot Delta vs. S(t)
ax = fig.add_subplot(3,1,3)
ax.scatter(S1, delta1, c='r', marker='x')
ax.scatter(S2, delta2, c='b', marker='x')

# Linear regression
b, m = np.polynomial.polynomial.polyfit(S1, delta1, 1)
ax.plot(S1, b + m * S1, '-', c='r', linewidth=0.5)
b, m = np.polynomial.polynomial.polyfit(S2, delta2, 1)
ax.plot(S2, b + m * S2, '-', c='b', linewidth=0.5)

ax.set_title('Dependence of Delta on Stock Price')
ax.set_xlabel('Stock Price')
ax.set_ylabel('Delta')
ax.legend(('Delta 1', 'Delta 2'))

fig.tight_layout()
plt.show()

if save_figure:
    fig.savefig('SFEDeltaHedgingLogic.png', transparent=True)
